pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -78,36 +76,41 @@ def get_profiled_train_dataloader(self, train_dataloader) -> Iterable:
     def prepare_data(self) -> None:
         # on multi-gpu jobs we only want to manipulate (download, etc) on node_rank=0, local_rank=0
         # or in the case where each node needs to do its own manipulation in which case just local_rank=0
-        should_call_once_per_node = self.trainer.local_rank == 0
-        should_call_once_in_total = self.trainer.local_rank == 0 and self.trainer.node_rank == 0
+        local_rank_zero = self.trainer.local_rank == 0
+        global_rank_zero = self.trainer.local_rank == 0 and self.trainer.node_rank == 0
 
         # handle datamodule prepare data:
         # check for prepare_data_per_node & datamodule lifecycle properties before calling datamodule.prepare_data
-        if self.trainer.datamodule is not None:
-            if is_overridden("prepare_data", self.trainer.datamodule) and (
-                not self.trainer.datamodule.has_prepared_data
+        if self.trainer.datamodule is not None and (not self.trainer.datamodule.has_prepared_data):
+
+            dm_prepare_data_per_node = self.trainer.datamodule.prepare_data_per_node
+            if (self.trainer.prepare_data_per_node is not None) and (
+                dm_prepare_data_per_node != self.trainer.prepare_data_per_node
             ):
-                dm_prepare_data_per_node = (
-                    self.trainer.prepare_data_per_node
-                    if self.trainer.datamodule is None
-                    else self.trainer.datamodule.prepare_data_per_node
+                raise MisconfigurationException(
+                    f"`prepare_data_per_node` should not be set in Trainer, got {self.trainer.prepare_data_per_node}. "
+                    f"`self.trainer.prepare_data_per_node`(={self.trainer.prepare_data_per_node}) is inconsistent with "
+                    f"`self.trainer.datamodule.prepare_data_per_node`"
+                    f"(={self.trainer.datamodule.prepare_data_per_node})."
                 )
```

Comment:
the error messages shouldn't reference `self.trainer.` as these are implementation details

we could say something along the lines of
```
Inconsistent settings found for `prepare_data_per_node`. Value was set with both `Trainer(prepare_data_per_node={}.)` and `DataModule.prepare_data_per_node={}`. Move `prepare_data_per_node` setting to DataModule property
```